src/client/connstring.rs

```diff
@@ -0,0 +1,318 @@
+        ConnectionOptions {
+            options: options,
+            read_pref_tags: read_pref_tags,
+        }
+    }
+
+    /// Retrieves an option using a borrowed String
+    pub fn get(&self, key: &String) -> Option<&String> {
+        self.options.get(key)
+    }
+
+    /// Retrieves an option using a borrowed str
+    pub fn get_str(&self, key: &str) -> Option<&str> {
+        match self.options.get(&key.to_owned()) {
+            Some(val) => Some(&val),
+            None => None,
+        }
+    }
+}
+
+/// Encapsulates information for connection to a single MongoDB host or replicated set.
+pub struct ConnectionString {
+    ctype: ConnectionType,
+    pub hosts: Vec<Host>,
+    pub string: Option<String>,
+    pub user: Option<String>,
+    pub password: Option<String>,
+    pub database: Option<String>,
+    pub collection: Option<String>,
+    pub options: Option<ConnectionOptions>,
+}
+
+impl ConnectionString {
+    /// Creates a new ConnectionString for a single, unreplicated host.
+    pub fn new(host_name: &str, port: u16) -> ConnectionString {
+        let host = Host::new(host_name.to_owned(), port);
+        ConnectionString::new_from_host(host)
+    }
+
+    fn new_from_host(host: Host) -> ConnectionString {
+        ConnectionString {
+            ctype: ConnectionType::Master,
+            hosts: vec![host],
+            string: None,
+            user: None,
+            password: None,
+            database: Some("test".to_owned()),
+            collection: None,
+            options: None,
+        }
+    }
+
+    /// Returns the connection type as an owned string.
+    pub fn type_to_string(ctype: ConnectionType) -> String {
+        let string = match ctype {
+            ConnectionType::Invalid => "invalid",
+            ConnectionType::Master => "master",
+            ConnectionType::Pair => "pair",
+            ConnectionType::Set => "set",
+        };
+        string.to_owned()
+    }
+}
+
+/// Parses a MongoDB connection string URI as defined by
+/// [the manual](http://docs.mongodb.org/manual/reference/connection-string/).
+pub fn parse(address: &str) -> Result<ConnectionString, &str> {
+    if !address.starts_with(URI_SCHEME) {
+        return Err("MongoDB connection string must start with 'mongodb://'")
+    }
+
+    // Remove scheme
+    let addr = &address[URI_SCHEME.len()..];
+
+    let mut ctype: ConnectionType;
+    let mut hosts: Vec<Host>;
+    let mut user: Option<String> = None;
+    let mut password: Option<String> = None;
+    let mut database: Option<String> = Some("test".to_owned());
+    let mut collection: Option<String> = None;
+    let mut options: Option<ConnectionOptions> = None;
+
+    // Split on host/path
+    let (host_str, path_str) = match addr.contains(".sock") {
+        true => {
+            let (host_part, path_part) = rpartition(addr, "/");
+            let host_test_uri = &format!("{}{}", URI_SCHEME, host_part);
```

Comment:
Don't really like the extra string creation here, we should refactor this so that there's an internal function that parses connection string minus scheme if anything